### PR TITLE
Stricter checks for typedesc

### LINKED ISSUE #8610: `const Foo=int` compiles; is that legal? what does it mean?

### LINKED ISSUE #8654: Compiler crash with `typedesc` as variable
Don't think this should work, though.
```nim
var a: typedesc
a = typedesc[int]
echo a is typedesc[int]
```
Error
```
Error: internal error: getTypeDescAux(tyNone)
```
And
```nim
const a: typedesc = typedesc[int]
echo a is typedesc[int]
```
Error
```
Error: internal error: genMagicExpr: mIs
```

### PR DESCRIPTION
**Pull request by hlaaftana**
Fixes #8610, fixes #8654

### CODE REVIEW

**Comment by mratsim** [comment]
Can't we just put an error message and that says:

```Nim
const Foo=int # Error: use "type Foo = int" instead
```

Special casing `const` for typedesc is a bad idea.

In the VM, `type` and NimNode should cover all use cases already.

File: compiler/types.nim [hunk 1]
```diff
@@ -1183,8 +1183,8 @@ proc typeAllowedAux(marker: var IntSet, typ: PType, kind: TSymKind,
     if result.isNil and t.sons[0] != nil:
       result = typeAllowedAux(marker, t.sons[0], skResult, flags)
   of tyTypeDesc:
-    # XXX: This is still a horrible idea...
-    result = nil
+    # Could be inside a macro, so var and let are allowed here and checked later
+    if kind notin {skParam, skResult, skVar, skLet}: result = t
```
Comment (hlaaftana) [inline compiler/types.nim]: Someone might know how to fix the concept tests without reverting this, so I'm keeping it for now unless someone thinks it should be reverted or if it doesn't get merged

File: tests/errmsgs/t8654_2.nim [hunk 2]
```diff
@@ -0,0 +1,7 @@
+discard """
+  errormsg: "cannot have typedesc as const value, use 'type a = int' instead"
+  line: 6
+"""
+
+const a: typedesc = typedesc[int]
+echo a is typedesc[int]
```
Comment (timotheecour) [inline tests/errmsgs/t8654_2.nim]: if you're gonna error on line 6, why have line 7? (+ same in other files)

Comment (hlaaftana) [reply tests/errmsgs/t8654_2.nim]: Sanity check? I copied existing tests

Comment (timotheecour) [reply tests/errmsgs/t8654_2.nim]: why is that a sanity check, if it won't even get compiled?
seems like you could put whatever there instead...
maybe existing tests could be updated too

**Comment by hlaaftana** [comment]
I'm not eligible to fix the test failures, so I'm gonna close the PR and not abandon it and make everyone wait forever, if anyone wants to take over they can start another PR

**Pull request closed without merging**